ic-stable-structures/src/memory_mapped_files/memory_mapped_file.rs

```diff
@@ -0,0 +1,358 @@
+            file,
+            path,
+            is_permanent,
+            length,
+            mapping,
+        })
+    }
+
+    /// Returns the current length in bytes
+    pub fn len(&self) -> u64 {
+        self.length
+    }
+
+    /// Resize the memory mapped file.
+    /// `new_length` should be `PAGE_SIZE` multiple.
+    /// If `new_length` is less or equal than the current length
+    /// nothing happens.
+    pub fn resize(&mut self, new_length: u64) -> MemMapResult<u64> {
+        if new_length % PAGE_SIZE != 0 {
+            return Err(MemMapError::SizeShouldBePageSizeMultiple);
+        }
+
+        if new_length < self.length {
+            return Ok(self.length);
+        }
+
+        if new_length > MEM_MAP_RESERVED_LENGTH {
+            return Err(MemMapError::OutOfAddressSpace {
+                claimed: new_length,
+                limit: MEM_MAP_RESERVED_LENGTH as _,
+            });
+        }
+
+        // There is no need to remap after changing the size
+        self.file.set_len(new_length)?;
+        self.length = new_length;
+
+        Ok(self.length)
+    }
+
+    /// Read data starting at `offset` to the given buffer.
+    pub fn read(&self, offset: u64, dst: &mut [u8]) -> MemMapResult<()> {
+        if offset + dst.len() as u64 > self.len() {
+            return Err(MemMapError::AccessOutOfBounds);
+        }
+
+        dst.copy_from_slice(&self.mapping[offset as usize..offset as usize + dst.len()]);
+
+        Ok(())
+    }
+
+    /// Write data from `src` to the memory starting at `offset`.
+    pub fn write(&mut self, offset: u64, src: &[u8]) -> MemMapResult<()> {
+        if offset + src.len() as u64 > self.len() {
+            return Err(MemMapError::AccessOutOfBounds);
+        }
+
+        self.mapping[offset as usize..offset as usize + src.len()].copy_from_slice(src);
+
+        Ok(())
+    }
+
+    /// Fill range with zeros.
+    pub fn zero_range(&mut self, offset: u64, count: u64) -> MemMapResult<()> {
+        if offset + count > self.length {
+            return Err(MemMapError::AccessOutOfBounds);
+        }
+
+        self.mapping[offset as _..(offset + count) as _].fill(0);
+
+        Ok(())
+    }
+
+    /// Flush all the changes to the underlying file.
+    pub fn flush(&self) -> MemMapResult<()> {
+        self.mapping.flush()?;
+
+        Ok(())
+    }
+
+    /// Save the copy to a file at the specified path.
+    pub fn save_copy(&self, path: impl AsRef<Path>) -> MemMapResult<()> {
+        self.flush()?;
+        copy(&self.path, path)?;
+
+        Ok(())
+    }
+
+    /// Set `is_permanent` flag.
+    pub fn set_is_permanent(&mut self, is_permanent: bool) {
+        self.is_permanent = is_permanent;
+    }
+}
+
+impl Drop for MemoryMappedFile {
+    fn drop(&mut self) {
+        if self.is_permanent {
+            self.flush().expect("failed to flush data to file")
+        } else {
+            _ = remove_file(&self.path);
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
could you please add some integration tests using this with a stable structure?

Reply:
I've added a test with filling stable structures and loading them from file